File: sorting_algorithms/max_heap_arr.py
from tools.decorators import statistics


def __parent_order(arr: list, index: int):

    while index > 0:
        parent_index = (index - 1) // 2
        if arr[parent_index] < arr[index]:
            arr[parent_index], arr[index] = arr[index], arr[parent_index]
        index = parent_index


def __max_heap_generator(arr: list, arr_len: int) -> list:

    for n in range(arr_len // 2):
        left = (n * 2) + 1
        right = (n * 2) + 2

        if right < arr_len:
            if arr[n] < arr[left] < arr[right]:
                arr[n], arr[right] = arr[right], arr[n]
            if arr[n] < arr[right] < arr[left]:
                arr[n], arr[left] = arr[left], arr[n]
            __parent_order(arr=arr, index=right)
        else:
            if arr[n] < arr[left]:
                arr[n], arr[left] = arr[left], arr[n]
        __parent_order(arr=arr, index=left)

    return arr


@statistics
def max_heap_sort(arr: list) -> list:

    arr_len = len(arr)
    __max_heap_generator(arr=arr, arr_len=arr_len)

    for n in range(arr_len - 1):
        last_position = arr_len - 1 - n
        arr[0], arr[last_position] = arr[last_position], arr[0]
        __max_heap_generator(arr=arr, arr_len=last_position)

    return arr

File: sorting_algorithms/max_heap_arr.js
const statistics = require("../tools/decorators")

/**
 * Get the index of the max value between parent and boot children
 * @param {Array} arr Array to sort
 * @param {Number} index index of the array to compare
 * @returns {Number} max index value, "-1" if parent have the max value
 */
const maxIndexValue = (arr, index) => {
  const left = index * 2 + 1;
  const right = index * 2 + 2;

  let max = left;
  if (arr[left] < arr[right]) {
    max = right;
  }
  if (arr[index] > arr[max]) {
    max = -1;
  }

  return max;
}

const maxHeap = (arr) => {
  // TODO, add as parameter length of array to dont have to create another array
  let breakFor = false;

  for (let index = 0; index < Math.floor(arr.length / 2); index++) {
    const maxIndex = maxIndexValue(arr, index);
    if (maxIndex >= 0) {
      [arr[maxIndex], arr[index]] = [arr[index], arr[maxIndex]];
      breakFor = true;
      break;
    }
  }
  if (breakFor) {
    maxHeap(arr);
  }
}


/**
 * Order an array use Max Heap Sort technique
 * @param {Array} arr array to sort
 * @returns {Boolean} if is ordered
 */
const maxHeapSort = (arr) => {
  maxHeap(arr);
  const orderedArr = [];
  const arrLength = arr.length - 1;

  for (let index = 0; index <= arrLength; index++) {
    const lastPosition = arrLength - index;
    [arr[lastPosition], arr[0]] = [arr[0], arr[lastPosition]]
    orderedArr.splice(0, 0, arr.pop());
    maxHeap(arr);
  }
  return orderedArr;
}

module.exports = { maxHeapSort: statistics(maxHeapSort) }
